module/VuFind/src/VuFind/Auth/LoginToken.php

```diff
@@ -0,0 +1,280 @@
+     * Create a new login token
+     *
+     * @param \VuFind\Db\Row\User $user      user
+     * @param string              $series    login token series
+     * @param string              $sessionId Session identifier
+     * @param int                 $expires   Token expiration date
+     *
+     * @return void
+     */
+    public function createToken(\VuFind\Db\Row\User $user, string $series = '', string $sessionId = '', $expires = 0)
+    {
+        $token = bin2hex(random_bytes(32));
+        $series = $series ? $series : bin2hex(random_bytes(32));
+        $browser = '';
+        $platform = '';
+        try {
+            $userInfo = get_browser(null, true) ?? [];
+            $browser = $userInfo['browser'];
+            $platform = $userInfo['platform'];
+        } catch (\Exception $e) {
+            // Problem with browscap.ini, continue without
+            // browser information
+        }
+        if ($expires === 0) {
+            $lifetime = $this->config->Authentication->persistent_login_lifetime ?? 0;
+            $expires = time() + $lifetime * 60 * 60 * 24;
+        }
+        $this->setLoginTokenCookie($user->id, $token, $series, $expires);
+
+        $this->loginTokenTable->saveToken($user->id, $token, $series, $browser, $platform, $expires, $sessionId);
+    }
+
+    /**
+     * Delete a login token by series. Also destroys
+     * sessions associated with the login token
+     *
+     * @param string $series Series to identify the token
+     * @param string $userId User identifier
+     *
+     * @return void
+     */
+    public function deleteTokenSeries(string $series, int $userId)
+    {
+        $cookie = $this->getLoginTokenCookie();
+        if (!empty($cookie) && $cookie['series'] === $series) {
+            $this->cookieManager->clear('loginToken');
+        }
+        if ($token = $this->loginTokenTable->getBySeries($series, $cookie['user_id'])) {
+            $handler = $this->sessionManager->getSaveHandler();
+            $handler->destroy($token->last_session_id);
+        }
+        $this->loginTokenTable->deleteBySeries($series, $cookie['user_id']);
+    }
+
+    /**
+     * Delete a login token from cookies and database
+     *
+     * @return void
+     */
+    public function deleteActiveToken()
+    {
+        $cookie = $this->getLoginTokenCookie();
+        if (!empty($cookie) && $cookie['series'] && $cookie['user_id']) {
+            $this->loginTokenTable->deleteBySeries($cookie['series'], $cookie['user_id']);
+        }
+        $this->cookieManager->clear('loginToken');
+    }
+
+    /**
+     * Send email warning to user
+     *
+     * @param User $user User
+     *
+     * @return void
+     */
+    public function sendLoginTokenWarningEmail(\VuFind\Db\Row\User $user)
+    {
+        if (!empty($user->email)) {
+            $this->mailer->send(
+                $user->email,
+                $this->config->Site->email,
+                $this->translate('login_warning_email_subject'),
```

Comment:
It might be a good idea to make the email subject line configurable via config.ini (but still translatable) so it's easier to change without modifying language files. It might also be worthwhile to make the email body a template so that can also be more easily customized. It might also be a good idea to include the site title in here somewhere, since otherwise with the default settings, the source of the email may not be immediately obvious to the end user if it isn't customized.